### PR TITLE
Add Code section to Contributions Doc

### PR DESCRIPTION
**Pull request by p-nayak11**
## Description
I transferred and editing the old contributing guide in robocup-software/doc. It has relevant, concise information with less confusing images. [Here](https://rj-rc-software.readthedocs.io/en/contributing_guidelines/) is the link to the docs.

### Test Case 1
1. Open link and check for typos.
2. Make sure content makes sense
3. Let me know if there is a section or pictures that I should add

### CODE REVIEW

**Comment by p-nayak11** [comment]
I feel that "add your branch as a version on the ReadTheDocs dashboard" part of Docs should be explained or written somewhere. Should this be a card or added to this PR? lmk

File: docs/source/contributing.rst [hunk 1]
```diff
@@ -1,12 +1,210 @@
...
+contributions as well as writing documentation. It is not intended to be a technical guide. Please refer to the tutorial docs for initial understanding of the technical skillset needed for RoboCup.
+
+If you don't have an overall idea of what the competition is like and what its
+goals are, you can brush up at the `RoboCup Wiki <http://wiki.robojackets.org/w/RoboCup>`_.
 
 Code
 ----
 
-TODO(Kevin): docstrings, GitHub, etc.
+To contribute to the codebase, you must know how to use Git, create Pull Requests, and document with docstrings. The following sections will give you some basic information about each of these, as well as addition resources to learn more about the topics.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
To contribute to the codebase, you must know how to use Git, create Pull Requests, and document with docstrings. The following sections will give you some basic information about each of these, as well as additional resources to learn more about the topics.
```

File: docs/source/contributing.rst [hunk 2]
```diff
@@ -1,12 +1,210 @@
...
+
+Git
+~~~
+
+We use Git as our version control system (if you already know git, you can skip
+ahead). A Version control system allows many people to code for the same project
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
ahead). A version control system allows many people to code for the same project
```

File: docs/source/contributing.rst [hunk 3]
```diff
@@ -1,12 +1,210 @@
...
+progress confused. You should use one (or several) peices of paper for each
+problem you are trying to solve, but should never use one sheet for multiple
+questions.
+
+.. figure:: https://git-scm.com/book/en/v2/images/basic-branching-2.png
+   :width: 80%
+   :align: center
+   :alt: Creating new branch
+
+   Creating new branch
```
Comment (kfu02) [inline docs/source/contributing.rst]: I don't love this image choice--it only shows the new iss53 branch pointing to the same commit as the master branch. I feel like any of these images from further down are more instructive:

Screenshot from 2022-06-17 11-18-20
Screenshot from 2022-06-17 11-18-07
Screenshot from 2022-06-17 11-17-53

Comment (kfu02) [reply docs/source/contributing.rst]: Also, the "C0" "C1" etc might be confusing to people who don't know what a "commit" is, can you work that into an explanation somewhere? It can be brief and you can reference the tutorial docs as somewhere members can go to learn more about git add/commit (since the start of those docs teach those basics)

File: docs/source/contributing.rst [hunk 4]
```diff
@@ -1,12 +1,210 @@
...
+your job to `resolve <https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line>`_ those errors. By merging changes into your contribution, rather than the other way around, you ensure the act of bringing your code into the team's repository will go smoothly. This helps when another member of the team reviews your code as well.
+
+Now that you have a merged branch, you should push the branch to your GitHub.
+From GitHub, you can make a `pull request <https://help.github.com/articles/using-pull-requests/>`_ from your repository against the team's repository.
+
+.. note::
+
+    A link to a edit a pull request is also automatically generated when the branch is pushed in the command line. You can use this link to create the pull request.
+
+Pull Requests(PR)
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
Pull Requests (PR)
```

File: docs/source/contributing.rst [hunk 5]
```diff
@@ -1,12 +1,210 @@
...
+~~~~~~~~~~~~~~~~~
+Pull requests are how we are able to review each others changes to the master branch. It helps you communicate your changes to ``ros2``. You can choose to create a pull request that is ready for review or draft a pull request. These draft PRs do not need to have reviewers assigned nor can it be merged into the master branch.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
Pull requests are how we are able to review each others changes to the master branch. It helps you communicate your changes to ``ros2``. You can choose to create a pull request that is ready for review or draft a pull request. Draft PRs must be marked as "ready for review" before they can be merged into the master branch.
```

File: docs/source/contributing.rst [hunk 6]
```diff
@@ -1,12 +1,210 @@
...
+
+Once your PR is ready for review, assign a team member to review the PR. This will notify the team member that you are ready to have your contribution reviewed. Requirements for pull request standards are listed in several sections below. The team member may ask that you fix or touch up some things before the request is accepted. This is normal and common. Once the pull request meets standards, the older member will approve it, and you changes will be complete.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
Once your PR is ready for review, assign an older member to review the PR. This will notify them that you are ready to have your contribution reviewed. Requirements for pull request standards are below. The team member may ask that you fix or touch up some things before the request is accepted. This is normal and common. Once the pull request meets standards, the older member will approve it, and your changes can be merged (provided they pass the CI checks, which is explained below).
```

File: docs/source/contributing.rst [hunk 7]
```diff
@@ -1,12 +1,210 @@
...
+
+Keep in mind, you can have several branches at once. If you need to fix a bug
+for an existing contribution while working on a new one, you should checkout
+the master branch, and then create a new branch named bug fix. It is
+critically important that a pull request only address one thing at a time. If it
+does not, the request will not accepted until you have separated the items you
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
does not, the request will not be accepted until you have properly separated the items you
```

File: docs/source/contributing.rst [hunk 8]
```diff
@@ -1,12 +1,210 @@
...
+have worked on.
+
+**Pull Request Requirements**
+
+With large projects, organization and structure can break down fairly quickly
+resulting in some bad spaghetti code. We want to avoid this as much as
+possible. For this reason, every pull request must be reviewed with patience. Follow what the tests and descriptions say to test. It is imperitive that you check to see everything compiles and functions as described by the PR.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
possible. For this reason, every pull request must be reviewed with patience. Follow what the tests and descriptions say to test. It is imperative that you check to see everything compiles and functions as described by the PR.
```

File: docs/source/contributing.rst [hunk 9]
```diff
@@ -1,12 +1,210 @@
...
 
-Python docstrings: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html
+If your request isn't accepted right away, don't take it personally. Often your code may work fine, but there are things you can add or refine. This helps keep our repository clean and will give you valuable experience participating in a code review process.
+
+**Continuous Integration**
+
+Continuous Integration (CI) is a tool to help auto-detect problems before they
+are merged into the main respository and have a chance to cause problems. Every
+time you submit a pull request, the CI tool is run and it will assign a passing
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
time you submit a pull request, our CI tools run and will assign a passing
```
(there are technically multiple CI tools)

File: docs/source/contributing.rst [hunk 10]
```diff
@@ -1,12 +1,210 @@
...
+or failing mark to the request. Currently, the CI checks consists of  build-and-test, mypy, pylint, and style. If the CI fails, you will need to fix the error in your code before the code review.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
or failing mark to the request. If the CI fails, you will need to fix the error in your code before the code review.
```
Just a personal preference thing, but I think making the docs as general as possible means we don't have to update them every year.

File: docs/source/contributing.rst [hunk 11]
```diff
@@ -1,12 +1,210 @@
...
+
+.. note::
+
+    At the time of this document's creation, the mypy and pylint errors are being ignored. However, you must pass the other two checks.
```
Comment (kfu02) [inline docs/source/contributing.rst]: ^ comment above

File: docs/source/contributing.rst [hunk 12]
```diff
@@ -1,12 +1,210 @@
...
+
+If the reason for the failure isn't obvious or it's a problem in the CI check itself, seek some help.
+
+**Content**
+
+It's important that the content of a pull request be kept clean and small. Pull
+requests should be less than 1-2k lines of code. The code changes should
+reflect one and only one topic (e.g do not include two bug fixes in one pull
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
reflect one and only one topic (e.g. do not include two bug fixes in one pull
```

File: docs/source/contributing.rst [hunk 13]
```diff
@@ -1,12 +1,210 @@
...
+request). Content should generally be kept to code and documentation, binary
+content, such as images, may be uploaded elsewhere.
+
+The description of the PR should accurately describe what the purpose of the changes are. The associated issue should link any PR or TODOs that are being addressed. A large refactor will require design documents to be written. Steps to test is a bulleted list that helps your reviewers on how to test your PR.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
The description of the PR should accurately describe what the purpose of the changes are. The associated issue should link any PR or TODOs that are being addressed. A large refactor will require design documents to be written. Steps to test should be a bulleted list that tells your reviewers how to test your PR, and what the expected result should be.
```

File: rj_gameplay/stp/formations/diamond_formation.py [hunk 14]
```diff
@@ -61,7 +61,7 @@ def get_regions(self) -> List:
     def get_centroids(self) -> List:
         """
         Convenience function for getting center of each region in get_regions()
-        :return: list of region centroids for X_formation
+        :return: list of region centroids for DiamondFormation
```
Comment (kfu02) [inline rj_gameplay/stp/formations/diamond_formation.py:64]: per your own contributing docs, these should be in a separate PR (but it's fine)

File: docs/source/contributing.rst [hunk 15]
```diff
@@ -1,12 +1,210 @@
...
+    :param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
+    :type [ParamName]: [ParamType](, optional)
+    ...
+    :raises [ErrorType]: [ErrorDescription]
+    ...
+    :return: [ReturnDescription]
+    :rtype: [ReturnType]
+    """
+
+Each function should have documentation containing it's purpose, what
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
Each function should have documentation containing its purpose, what
```

File: docs/source/contributing.rst [hunk 16]
```diff
@@ -1,12 +1,210 @@
...
+parameters is takes, and what values it returns. Error handling should be
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
parameters it takes, and what values it returns. Error handling should be
```

File: docs/source/contributing.rst [hunk 17]
```diff
@@ -1,12 +1,210 @@
...
+described as well: what errors will it produce, and what assumptions it makes
+regarding the validation the caller performs ahead of time.
+
+If the overall set of code is complex and new, consider editing or adding to
+the docs.
+
+For more information on writing Sphinx docstrings, you can learn more `here <https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html>`_.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
For more information on writing Sphinx docstrings, see `this link <https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html>`_.
```
I just find this wording a bit redundant

File: docs/source/contributing.rst [hunk 18]
```diff
@@ -1,12 +1,210 @@
...
+    :return: [ReturnDescription]
+    :rtype: [ReturnType]
+    """
+
+Each function should have documentation containing it's purpose, what
+parameters is takes, and what values it returns. Error handling should be
+described as well: what errors will it produce, and what assumptions it makes
+regarding the validation the caller performs ahead of time.
+
+If the overall set of code is complex and new, consider editing or adding to
```
Comment (kfu02) [inline docs/source/contributing.rst]: What does this mean?

File: docs/source/contributing.rst [hunk 19]
```diff
@@ -1,12 +1,210 @@
...
+the docs.
+
+For more information on writing Sphinx docstrings, you can learn more `here <https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html>`_.
+
+Text file types that are not supported by doxygen or Sphinx should still be documented using what ever commenting style that format supports.
```
Comment (kfu02) [inline docs/source/contributing.rst]: What does this mean? I assume it's referring to non-Python code but it's a little unclear what I should do if I have C++ code changes

File: docs/source/contributing.rst [hunk 20]
```diff
@@ -1,12 +1,210 @@
...
+
+**Style and Formatting**
+
+In order to keep the code more readable, code should be formatted and styled
+uniformly. This would be difficult to coordinate across multiple users, so we
+have a program that automatically restyles the code for you. Our precommit hook runs ``make pretty-lines`` which auto-formats the code.
```
Comment (kfu02) [inline docs/source/contributing.rst]: actually right now pretty-lines precommit hook is a little broken, can you add that as a card

File: docs/source/contributing.rst [hunk 21]
```diff
@@ -1,12 +1,210 @@
...
+
+.. note::
+
+    You may have to readd the foramtting changes, and commit again.
+
+You can check if the style is passing by running ``make checkstyle``. If there are no errors, then you are good to go.
+
+If you've done all this successfully, you are now an official contributor.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
If you've done all this successfully, you are now an official contributor. Merge your PR!
```
just some extra positivity

File: docs/source/contributing.rst [hunk 22]
```diff
@@ -28,9 +226,13 @@ it in the toctree in ``docs/source/index.rst``. (Toctree = table of contents
...
 
 You can check your work by asking whoever maintains these docs (as of 2022,
-Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard. They
-should go to the webpage, add a version that points to the new doc development
-branch, and build that version, then give you the link.
+Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard.
+
+.. note::
+
+    If you are maintaining these docs, you must install the requirements in ``docs/requirements.txt``.
```
Comment (kfu02) [inline docs/source/contributing.rst]: you actually don't need to unless you're building locally with `make docs`, can you add a new section that explains local building?

**Review by kfu02 — CHANGES_REQUESTED** [review]
Good content but typos were distracting. I like to do a proofreading pass with the Files Changed tab before requesting reviews, maybe something to consider.

There's also very little on the cpp side for docstrings (which is understandable since you haven't touched that side yet). In general:
 * docstrings should go in associated .hpp files (header = interface so docstrings are useful there)
 * smaller comments should go in source
 * ~add the [Breathe formatting for C++](https://breathe.readthedocs.io/en/latest/) here too (even if we don't conform to it rn)~

I'm trying to follow these rules in #1908 if you need an example.

EDIT: the last bullet should be added in a different PR that also enables autodoc for C++. So I'm scratching it out for now

File: docs/source/contributing.rst [hunk 23]
```diff
@@ -1,12 +1,223 @@
...
+
+The guide for C++ docstrings is followed by most C++ files in our codebase. Include the doctrings in the ``.hpp`` files.
+
+The following is the docstring format:
+
+.. code-block:: python
+
+    """
+    /**
+     * @brief [Breif Summary of Method]
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
     * @brief [Brief Summary of Method]
```

File: docs/source/contributing.rst [hunk 24]
```diff
@@ -19,18 +230,31 @@ Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``
...
-Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard. They
-should go to the webpage, add a version that points to the new doc development
-branch, and build that version, then give you the link.
+You can check your work by building locally or asking whoever maintains these docs (as of 2022, Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard.
+
+**Local Building**
+
+Locally building the docs is a troublesome process to check your branch's changes. Unless you are the person who maintains the docs, it is not worth the trouble. However, this section explains how to locally build our docs.
+
+You must install all of the dependencies in ``docs/requirements.txt`` file. After doing so, change directories to the ``docs`` folder in our code base and ``make html``. This will create multiple build files that should NOT be commited to your branch nor ``ros2``.
```
Comment (kfu02) [inline docs/source/contributing.rst]: this can be done with 
`python3 -m pip install -r requirements.txt`

FYI

File: docs/source/contributing.rst [hunk 25]
```diff
@@ -19,18 +230,31 @@ Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``
...
+
+.. note::
+
+    DO NOT COMMIT THE BUILD FILES GENERATED WHEN LOCALLY BUILDING!
+
+This will generate the docs with the changes made from your branch. When commiting to your branch, make sure to exclude the build files.
+
+**Branch Version**
+
+The dcoumentation maintainer should go to the webpage, activate a version that points to the new doc development branch, and build that version, then give you the link.
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
The documentation maintainer should go to the ReadTheDocs dashboard for our repo, activate the version that points to the new doc development branch, then give you the link once it builds.
```

File: docs/source/contributing.rst [hunk 26]
```diff
@@ -19,18 +230,31 @@ Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``
...
 
-You can check your work by asking whoever maintains these docs (as of 2022,
-Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard. They
-should go to the webpage, add a version that points to the new doc development
-branch, and build that version, then give you the link.
+You can check your work by building locally or asking whoever maintains these docs (as of 2022, Kevin Fu) to add your branch as a version on the ReadTheDocs dashboard.
+
+**Local Building**
+
+Locally building the docs is a troublesome process to check your branch's changes. Unless you are the person who maintains the docs, it is not worth the trouble. However, this section explains how to locally build our docs.
```
Comment (kfu02) [inline docs/source/contributing.rst]: this is a little lengthy, you could sum it up as
```suggestion
This section explains how to locally build our docs. Locally building the docs is not worth the trouble unless you don't have access to the RTD dashboard and whoever does can't be reached. 
```

Comment (kfu02) [reply docs/source/contributing.rst]: I would also move the section below on building via RTD dashboard above this, so people read that first

File: docs/source/contributing.rst [hunk 27]
```diff
@@ -19,18 +230,31 @@ Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``
...
+
+You must install all of the dependencies in ``docs/requirements.txt`` file. After doing so, change directories to the ``docs`` folder in our code base and ``make html``. This will create multiple build files that should NOT be commited to your branch nor ``ros2``.
+
+.. note::
+
+    DO NOT COMMIT THE BUILD FILES GENERATED WHEN LOCALLY BUILDING!
+
+This will generate the docs with the changes made from your branch. When commiting to your branch, make sure to exclude the build files.
+
+**Branch Version**
```
Comment (kfu02) [inline docs/source/contributing.rst]:
```suggestion
**Building via RTD**
```

**Review by kfu02 — CHANGES_REQUESTED** [review]
Contributing guide looks awesome! Left a few requested changes on the doc-building side

File: docs/source/contributing.rst [hunk 28]
```diff
@@ -1,12 +1,223 @@
...
+You now have a branch with your contribution, but you haven't contributed until
+the code makes it into the main repository. This involves several steps. First,
+any changes others have made in the team's repository need to be merged into
+your code.
+
+.. tip::
+
+    If the work done on the branch stretches over a long period, it is good practice to frequently merge the master branch into your personal branch.
+
+If there are any conflics Git can't resolve automatically, it is
```
Comment (kasohrab) [inline docs/source/contributing.rst]:
```suggestion
If there are any conflicts that Git can't resolve automatically, it is
```

File: docs/source/contributing.rst [hunk 29]
```diff
@@ -19,18 +230,31 @@ Our docs are hosted by ReadTheDocs and are built from ReStructuredText (``.rst``
...
+
+**Local Building**
+
+Locally building the docs is a troublesome process to check your branch's changes. Unless you are the person who maintains the docs, it is not worth the trouble. However, this section explains how to locally build our docs.
+
+You must install all of the dependencies in ``docs/requirements.txt`` file. After doing so, change directories to the ``docs`` folder in our code base and ``make html``. This will create multiple build files that should NOT be commited to your branch nor ``ros2``.
+
+.. note::
+
+    DO NOT COMMIT THE BUILD FILES GENERATED WHEN LOCALLY BUILDING!
```
Comment (kasohrab) [inline docs/source/contributing.rst]: Shouldn't the .gitignore take care of this?

File: docs/source/contributing.rst [hunk 30]
```diff
@@ -1,12 +1,223 @@
...
+
+If there are any conflics Git can't resolve automatically, it is
+your job to `resolve <https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line>`_ those errors. By merging changes into your contribution, rather than the other way around, you ensure the act of bringing your code into the team's repository will go smoothly. This helps when another member of the team reviews your code as well.
+
+Now that you have a merged branch, you should push the branch to your GitHub.
+From GitHub, you can make a `pull request <https://help.github.com/articles/using-pull-requests/>`_ from your repository against the team's repository.
+
+.. note::
+
+    A link to a edit a pull request is also automatically generated when the branch is pushed in the command line. You can use this link to create the pull request.
```
Comment (kasohrab) [inline docs/source/contributing.rst]:
```suggestion
    A link to a edit a pull request is also automatically generated when the branch is pushed via the command line. You can use this link to create the pull request.
```

File: docs/source/contributing.rst [hunk 31]
```diff
@@ -1,12 +1,223 @@
...
+
+Once your PR is ready for review, assign an older team member to review the PR. This will notify them that you are ready to have your contribution reviewed. Requirements for pull request standards are listed in several sections below. The team member may ask that you fix or touch up some things before the request is accepted. This is normal and common. Once the pull request meets standards, the older member will approve it, and your changes can be merged (provided they pass the CI checks, which are explained below).
+
+Keep in mind, you can have several branches at once. If you need to fix a bug
+for an existing contribution while working on a new one, you should checkout
+the master branch, and then create a new branch named bug fix. It is
+critically important that a pull request only addresses one thing at a time. If it does not, the request will not be accepted until you have properly separated the items you have worked on.
```
Comment (kasohrab) [inline docs/source/contributing.rst:119]: If you want to enforce this, updating out PR templates would help with that. [here ](https://github.com/UBC-Thunderbots/Software/blob/master/.github/PULL_REQUEST_TEMPLATE.md)is a nice example of a template from Thunderbots

Comment (kasohrab) [reply docs/source/contributing.rst:119]: They haven't been updated in 2 years, so I think it is time to refresh them.

**Review by kasohrab — CHANGES_REQUESTED** [review]
great work! I left some suggestions.

**Review by kasohrab — APPROVED** [review]
lgtm

**Review by kfu02 — APPROVED** [review]
LGTM! now the challenge is enforcing all those docstring rules...
